src/Cli/dotnet/SlnFileFactory.cs

```diff
@@ -19,87 +49,57 @@ public static string[] ListSolutionFilesInDirectory(string directory, bool inclu
-            }
-            catch (SolutionException e)
-            {
-                throw new GracefulException(
-                    CommonLocalizableStrings.InvalidSolutionFormatString,
-                    solutionPath,
-                    e.Message);
-            }
-            return slnFile;
+            return serializer.OpenAsync(solutionPath, CancellationToken.None).Result;
         }
 
-        private static SolutionModel FromDirectory(string solutionDirectory, bool includeSolutionXmlFiles)
+        public static SolutionModel CreateFromFilteredSolutionFile(string filteredSolutionFile)
         {
-            DirectoryInfo dir;
-            try
-            {
-                dir = new DirectoryInfo(solutionDirectory);
-                if (!dir.Exists)
-                {
-                    throw new GracefulException(
-                        CommonLocalizableStrings.CouldNotFindSolutionOrDirectory,
-                        solutionDirectory);
-                }
-            }
-            catch (ArgumentException)
-            {
-                throw new GracefulException(
-                    CommonLocalizableStrings.CouldNotFindSolutionOrDirectory,
-                    solutionDirectory);
-            }
-
-            FileInfo[] files = [
-                ..dir.GetFiles("*.sln"),
-                ..(includeSolutionXmlFiles ? dir.GetFiles(".slnx") : [])
-             ];
+            JsonDocument jsonDocument = JsonDocument.Parse(File.ReadAllText(filteredSolutionFile));
+            JsonElement jsonElement = jsonDocument.RootElement;
+            JsonElement filteredSolutionJsonElement = jsonElement.GetProperty("solution");
+            string originalSolutionPath = filteredSolutionJsonElement.GetProperty("path").GetString();
+            string originalSolutionPathAbsolute = Path.GetFullPath(Path.GetDirectoryName(originalSolutionPath), filteredSolutionFile);
+            string[] filteredSolutionProjectPaths = filteredSolutionJsonElement.GetProperty("projects")
+                .EnumerateArray()
+                .Select(project => project.GetString())
+                .ToArray();
 
-            if (files.Length == 0)
+            if (string.IsNullOrEmpty(originalSolutionPathAbsolute) || !File.Exists(originalSolutionPathAbsolute))
             {
                 throw new GracefulException(
-                    CommonLocalizableStrings.CouldNotFindSolutionIn,
-                    solutionDirectory);
+                    CommonLocalizableStrings.InvalidSolutionFormatString,
+                    filteredSolutionFile, "");
             }
 
-            if (files.Length > 1)
-            {
-                throw new GracefulException(
-                    CommonLocalizableStrings.MoreThanOneSolutionInDirectory,
-                    solutionDirectory);
-            }
+            SolutionModel filteredSolution = new SolutionModel();
+            SolutionModel originalSolution = CreateFromFileOrDirectory(originalSolutionPath);
```

Comment:
Do you need the absolute one?

Reply:
Yes, as that makes sure the path is relative to the solution file, and not the working directory. It is an edge case but could still happen

Reply:
I meant, do you need originalSolutionPathAbsolute?

Reply:
Oh you're absolutely right, I missed it. Thanks!